File: 2_valid_parentheses_20.py
'''
https://leetcode.com/problems/valid-parentheses/description/
'''

# 100% result. 

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        values_dict = {
            "}": "{",
            ")": "(",
            "]": "["
        }

        for elem in s:
            
            if elem in values_dict:

                if stack and stack[-1] == values_dict[elem]:
                    stack.pop()
                    continue

            stack.append(elem)

        return not stack


print(Solution().isValid("()[]{}"))

print(Solution().isValid("(]"))

File: 2_valid_parentheses_20.go
// https://leetcode.com/problems/valid-parentheses/description/

package main

import "fmt"

// 100% solution

func isValid(s string) bool {
	valuesDict := map[rune]rune{
		'}': '{',
		')': '(',
		']': '[',
	}

	var stack []rune

	for _, char := range s {

		if opening, found := valuesDict[char]; found {

			if len(stack) > 0 && stack[len(stack)-1] == opening {
				stack = stack[:len(stack)-1]
				continue
			}
		}
		stack = append(stack, char)
	}

	return len(stack) == 0
}

func main() {
	fmt.Println(isValid("()[]{}")) // true
	fmt.Println(isValid("(]"))     // false
}
